synthetics-stats/src/entities/volume.ts

```diff
@@ -0,0 +1,119 @@
+  }
+
+  hourlyVolumeInfo.save();
+  dailyVolumeInfo.save();
+  totalVolumeInfo.save();
+}
+
+function getOrCreateVolumeInfo(timestamp: i32, period: string): VolumeInfo {
+  let timestampGroup = timestampToPeriodStart(timestamp, period);
+  let volumeId = period === "total" ? "total" : timestampGroup.toString();
+  let volumeInfo = VolumeInfo.load(volumeId);
+
+  if (volumeInfo === null) {
+    volumeInfo = new VolumeInfo(volumeId);
+    volumeInfo.period = period;
+    volumeInfo.volumeUsd = BigInt.fromI32(0);
+    volumeInfo.swapVolumeUsd = BigInt.fromI32(0);
+    volumeInfo.marginVolumeUsd = BigInt.fromI32(0);
+  }
+  return volumeInfo as VolumeInfo;
+}
+
+export function saveSwapVolumeInfo(eventData: EventData, transaction: Transaction): void {
+  let PERIODS = new Array<string>(3);
+  PERIODS[0] = "1h";
+  PERIODS[1] = "1d";
+  PERIODS[2] = "total";
+
+  let timestamp = transaction.timestamp;
+  let tokenIn = eventData.getAddressItemString("tokenIn")!;
+  let tokenOut = eventData.getAddressItemString("tokenOut")!;
+  let amountIn = eventData.getUintItem("amountIn")!;
+  let tokenInPrice = eventData.getUintItem("tokenInPrice")!;
+  let volumeUsd = amountIn.times(tokenInPrice);
+
+  createOrUpdateSwapVolumeInfo(PERIODS, timestamp, volumeUsd, tokenIn, tokenOut);
+}
+
+function createOrUpdateSwapVolumeInfo(periodsToCreate: string[], timestamp: i32, volumeUsd: BigInt, tokenIn: string, tokenOut: string): void {
+  for (let i = 0; i < periodsToCreate.length; i++) {
+    let period = periodsToCreate[i];
+    let timestampGroup = timestampToPeriodStart(timestamp, period);
+    let id = period === "total" ? "total" : getVolumeInfoByTokenId(tokenIn, tokenOut) + ":" + timestampGroup.toString();
+    let volumeInfo = SwapVolumeInfo.load(id);
+
+    if (volumeInfo === null) {
+      volumeInfo = new SwapVolumeInfo(id);
+      volumeInfo.tokenIn = tokenIn;
+      volumeInfo.tokenOut = tokenOut;
+      volumeInfo.timestamp =timestampGroup;
+      volumeInfo.period = period;
+      volumeInfo.volumeUsd = BigInt.fromI32(0);
+    }
+
+    volumeInfo.volumeUsd = volumeInfo.volumeUsd.plus(volumeUsd);
+    volumeInfo.save();
+  }
+}
+
+export function savePositionVolumeInfo(eventData: EventData, transaction: Transaction): void {
+  let PERIODS = new Array<string>(3);
+  PERIODS[0] = "1h";
+  PERIODS[1] = "1d";
+  PERIODS[2] = "total";
+  let timestamp = transaction.timestamp;
+  let collateralToken = eventData.getAddressItemString("collateralToken")!;
+  let marketToken = eventData.getAddressItemString("market")!;
+  let marketInfo = MarketInfo.load(marketToken)!;
+  let sizeInUsd = eventData.getUintItem("sizeInUsd")!;
+
+  createOrUpdatePositionVolumeInfo(PERIODS, timestamp, sizeInUsd, collateralToken, marketInfo.indexToken);
+}
+
+function createOrUpdatePositionVolumeInfo(periodsToCreate: string[], timestamp: i32, volumeUsd: BigInt, collateralToken: string, indexToken: string): void {
```

Comment:
this feels a bit complex because one function iterates on periods, creates entities and updates the fields

maybe refactor it to just accept one period and just call the function 3 times in savePositionValumeInfo or in the mapping
you can find example in saveCollectedMarketFeesForPeriod which called in mapping.ts for 3 periods

Reply:
savePositionVolumeInfo(..., '1h');
savePositionVolumeInfo(..., '1d');
savePositionVolumeInfo(..., 'total')

Reply:
Done!